### PR TITLE
PR: Fix Duplicate Metric Logging in MLFlowLogger to Prevent MLflow Database Errors

### LINKED ISSUE #20865: Mlflow logging LR duplicate key issue with PostgreSQL DB #190
### Bug description

**What happened?**
In the anemoi codebase - we use the pytorch lightning MlflowLogger and we hit this error when logging this to a remote server with a postgresql database:

`mlflow.exceptions.RestException: BAD_REQUEST: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "metric_pk"
DETAIL:  Key (key, "timestamp", step, run_uuid, value, is_nan)=(lr-AdamW, 1741862947817, 60899, 843c1331fefa436bab56485cfc5bc16e, 0.000322870392593084, f) already exists.`

https://github.com/ecmwf/anemoi-core/blob/main/training/src/anemoi/training/diagnostics/mlflow/logger.py

versions:

```
mlflow                             2.22.0
pytorch-lightning           2.5.1
```

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by KAVYANSHTYAGI**
**What does this PR do?**

This PR fixes a long standing issue in PyTorch Lightning’s MLFlowLogger where logging the same metric (with the same name and step) more than once in a run causes a unique constraint violation on certain MLflow backends (e.g., PostgreSQL).
Now, MLFlowLogger tracks (metric, step) pairs and skips any duplicate metric logs within a run, preventing database errors and improving robustness.

This change also updates the class docstring to document this new behavior and adds a unit test to verify that duplicate metric logs are ignored as expected.

Fixes #20865

**Motivation and Context**

    Some MLflow tracking servers (such as those backed by PostgreSQL) enforce a unique constraint on metrics.

    If the same metric (with identical name and step) is logged more than once, MLflow returns an error and metric logging fails, potentially halting training.

    This situation often arises when users call .log() in multiple hooks or callbacks.

    The deduplication logic ensures only the first log of a metric per (name, step) is recorded per run.

Dependencies

    No new dependencies are introduced.

Does your PR introduce any breaking changes?

    No breaking changes .... existing behavior is preserved except that duplicate metric logs are now silently skipped (users may see a log message if a duplicate is skipped).

Other Checklist Items

Documentation updated- yes(see class docstring in MLFlowLogger)

New test added for deduplication- yes

Fun fact:
This change will help Lightning users avoid subtle training failures, especially with remote or production MLflow tracking servers!

### CODE REVIEW

File: tests/tests_pytorch/loggers/test_mlflow.py [hunk 1]
```diff
@@ -427,3 +427,50 @@ def test_set_tracking_uri(mlflow_mock):
     mlflow_mock.set_tracking_uri.assert_not_called()
     _ = logger.experiment
     mlflow_mock.set_tracking_uri.assert_called_with("the_tracking_uri")
+
+
+def test_mlflowlogger_metric_deduplication(monkeypatch):
+    import types
+
+    from lightning.pytorch.loggers.mlflow import MLFlowLogger
```
Comment (rittik9) [inline tests/tests_pytorch/loggers/test_mlflow.py:435]: lightning.pytorch.loggers import MLFlowLogger

**Comment by SkafteNicki** [comment]
will maybe also fix #20902

File: requirements/pytorch/test.txt [hunk 2]
```diff
@@ -18,3 +18,4 @@ fastapi  # for `ServableModuleValidator`  # not setting version as re-defined in
 uvicorn  # for `ServableModuleValidator`  # not setting version as re-defined in App
 
 tensorboard >=2.9.1, <2.21.0  # for `TensorBoardLogger`
+mlflow
```
Comment (SkafteNicki) [inline requirements/pytorch/test.txt]:
```suggestion
mlflow >=3.0.0, <4.0   # for `MLFlowLogger 
```

File: src/lightning/pytorch/loggers/mlflow.py [hunk 3]
```diff
@@ -109,6 +109,12 @@ def any_lightning_module_function_or_hook(self):
         ModuleNotFoundError:
             If required MLFlow package is not installed on the device.
 
+    Note:
+        As of vX.XX, MLFlowLogger will skip logging any metric (same name and step)
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loggers/mlflow.py]: No need to mention what version this was changed in
```suggestion
        MLFlowLogger will skip logging any metric (same name and step)
```

File: src/lightning/pytorch/loggers/mlflow.py [hunk 4]
```diff
@@ -109,6 +109,12 @@ def any_lightning_module_function_or_hook(self):
...
+        more than once per run, to prevent database unique constraint violations on
+        some MLflow backends (such as PostgreSQL). Only the first value for each (metric, step)
+        pair will be logged per run. This improves robustness for all users.
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loggers/mlflow.py]:
```suggestion
        pair will be logged per run.
```

File: src/lightning/pytorch/loggers/mlflow.py [hunk 5]
```diff
@@ -266,6 +274,12 @@ def log_metrics(self, metrics: Mapping[str, float], step: Optional[int] = None)
                     category=RuntimeWarning,
                 )
                 k = new_k
+
+            metric_id = (k, step or 0)
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loggers/mlflow.py:279]:
```suggestion
            # Prevent trying to log duplicate metric value
            metric_id = (k, step or 0)
```

File: src/lightning/pytorch/loggers/mlflow.py [hunk 6]
```diff
@@ -151,6 +157,7 @@ def __init__(
         from mlflow.tracking import MlflowClient
 
         self._mlflow_client = MlflowClient(tracking_uri)
+        self._logged_metrics = set()  # Track (key, step)
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loggers/mlflow.py]: Probably need to import `Set` and `Tuple` from typing
```suggestion
        self._logged_metrics: Set[Tuple[str, float]] = set()  # Track (key, step)
```

**Comment by SkafteNicki** [comment]
@KAVYANSHTYAGI could you please check failing tests?
